pkg/provisioner/conditionwatcher.go

```diff
@@ -0,0 +1,170 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package provisioner
+
+import (
+	"context"
+	"os"
+	"sync"
+	"time"
+
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/klog/v2"
+
+	"k8s.io/client-go/tools/cache"
+	azDiskClientSet "sigs.k8s.io/azuredisk-csi-driver/pkg/apis/client/clientset/versioned"
+	azurediskInformers "sigs.k8s.io/azuredisk-csi-driver/pkg/apis/client/informers/externalversions"
+)
+
+const (
+	azvolumeattachment = "azvolumeattachments"
+	azvolume           = "azvolume"
+	azdrivernode       = "azdrivernode"
+)
+
+type waitResult struct {
+	obj runtime.Object
+	err error
+}
+
+type waitEntry struct {
+	conditionFunc func(obj interface{}) (bool, error)
+	waitChan      chan waitResult
+}
+
+type conditionWatcher struct {
+	informerFactory azurediskInformers.SharedInformerFactory
+	waitMap         sync.Map // maps namespaced name to waitEntry
+	namespace       string
+}
+
+func newConditionWatcher(ctx context.Context, azDiskClient azDiskClientSet.Interface, namespace string, resync time.Duration) *conditionWatcher {
+	informerFactory := azurediskInformers.NewSharedInformerFactory(azDiskClient, resync)
+	azVolumeAttachmentInformer := informerFactory.Disk().V1alpha1().AzVolumeAttachments().Informer()
+	azVolumeInformer := informerFactory.Disk().V1alpha1().AzVolumes().Informer()
+	azDriverNodeInformer := informerFactory.Disk().V1alpha1().AzDriverNodes().Informer()
+
+	c := conditionWatcher{
+		informerFactory: informerFactory,
+		waitMap:         sync.Map{},
+	}
+
+	azVolumeAttachmentInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		UpdateFunc: c.onUpdate,
+		DeleteFunc: c.onDelete,
+	})
+	azVolumeInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		UpdateFunc: c.onUpdate,
+		DeleteFunc: c.onDelete,
+	})
+	azDriverNodeInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		UpdateFunc: c.onUpdate,
+		DeleteFunc: c.onDelete,
+	})
+
+	go informerFactory.Start(ctx.Done())
+
+	synced := cache.WaitForCacheSync(ctx.Done(), azVolumeAttachmentInformer.HasSynced, azVolumeInformer.HasSynced, azDriverNodeInformer.HasSynced)
+	if !synced {
+		klog.Fatalf("Unable to sync caches for azuredisk CRIs")
+		os.Exit(1)
+	}
+
+	return &c
+}
+
+func (c *conditionWatcher) waitForObjectCondition(ctx context.Context, objectType, objectName string, conditionFunc func(obj interface{}) (bool, error)) (*conditionWaiter, error) {
+	klog.V(5).Infof("Adding a condition function for %s (%s)", objectType, objectName)
+	entry := waitEntry{
+		conditionFunc: conditionFunc,
+		waitChan:      make(chan waitResult, 1),
+	}
+	_, exists := c.waitMap.LoadOrStore(objectName, &entry)
```

Comment:
Hmm. A map using the name as the key may be too simplistic as it doesn't handled collisions between types. Although we do not currently expect name collisions between types, this may be more likely in the future if we add more objects. Consider using a struct containing at least the type and object names for the key.

Reply:
Sounds good. Will do